pil-analyzer/src/type_inference.rs

```diff
@@ -60,28 +62,66 @@ impl From<Type> for ExpectedType {
     }
 }
 
-struct TypeChecker {
+#[derive(Debug, Clone)]
+struct DeclaredType {
+    pub source: SourceRef,
+    pub vars: TypeBounds,
+    pub ty: TypeDeclaredType,
+}
+
+impl DeclaredType {
+    fn scheme(&self) -> TypeScheme {
+        match &self.ty {
+            TypeDeclaredType::Struct(ty, _) => TypeScheme {
+                vars: self.vars.clone(),
+                ty: ty.clone(),
+            },
+            TypeDeclaredType::Type(ty) => TypeScheme {
+                vars: self.vars.clone(),
+                ty: ty.clone(),
+            },
+        }
+    }
+
+    fn type_mut(&mut self) -> &mut Type {
+        match &mut self.ty {
+            TypeDeclaredType::Struct(ty, _) => ty,
+            TypeDeclaredType::Type(ty) => ty,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+enum TypeDeclaredType {
```

Comment:
Hm, not the nicest name, but also don't know a better one :)

Reply:
mmm, what about `DeclaredTypeKind`?